### PR TITLE
run cpp, nosetests and python roslaunch in order

### PR DESCRIPTION
**Pull request by knorth55**
run cpp, nosetests, and python launch in order
this PR fix nosetests report.

### CODE REVIEW

File: cmake/Modules/CodeCoverage.cmake [hunk 1]
```diff
@@ -192,23 +192,26 @@ function(ADD_CODE_COVERAGE)
       set(OMIT_FLAGS --omit ${Coverage_EXCLUDES})
     endif()
 
+    add_custom_target(${Coverage_NAME}_nosetests_py
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:195]: No output?

File: cmake/Modules/CodeCoverage.cmake [hunk 2]
```diff
@@ -192,23 +192,26 @@ function(ADD_CODE_COVERAGE)
...
+        COMMAND ${PYTHON_COVERAGE_PATH} report --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python nosetests report to output"
+        COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_nosetests_python.xml --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python nosetests xml to output"
+        WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+        DEPENDS ${Coverage_NAME}_cpp
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:199]: Why do you need to run this command after cpp coverage test?

File: cmake/Modules/CodeCoverage.cmake [hunk 3]
```diff
@@ -192,23 +192,26 @@ function(ADD_CODE_COVERAGE)
...
-        COMMAND ${PYTHON_COVERAGE_PATH} combine || echo "WARNING: No python coverage to combine!"
-        COMMAND ${PYTHON_COVERAGE_PATH} report --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: no python report to output"
+        COMMAND ${PYTHON_COVERAGE_PATH} combine || echo "WARNING: No python coverage to combine"
+        COMMAND ${PYTHON_COVERAGE_PATH} report --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python report to output"
         COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_python.xml --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python xml to output"
-        COMMAND ${PYTHON_COVERAGE_PATH} html -d ${Coverage_NAME}_python_html --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python html to output"
-        COMMAND cp ${COVERAGE_DIR}/${Coverage_NAME}_python.xml ${PROJECT_BINARY_DIR} || echo "WARNING: No python xml to copy"
-        COMMAND cp -r ${COVERAGE_DIR}/${Coverage_NAME}_python_html ${PROJECT_BINARY_DIR} || echo "WARNING: No python html to copy"
+        COMMAND cp ${COVERAGE_DIR}/${Coverage_NAME}_python.xml ${PROJECT_BINARY_DIR}/ || echo "WARNING: No python xml"
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:208]: Don't we need to list this xml in OUTPUT?

File: cmake/Modules/CodeCoverage.cmake [hunk 4]
```diff
@@ -192,23 +192,26 @@ function(ADD_CODE_COVERAGE)
...
+        COMMAND cp ${COVERAGE_DIR}/.coverage* ${PROJECT_BINARY_DIR}/ || echo "WARNING: No python coverage"
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:209]: ditto

File: cmake/Modules/CodeCoverage.cmake (hunk 2 above)
Comment (knorth55) [reply cmake/Modules/CodeCoverage.cmake]: I'm worried about file crash between each coverage process.
nosetests and roslaunch python coverage have file name crash, so we need to run one by one.
But c++, maybe we don't need to do that.
I'm not pretty sure, so I move to output files to `cpp_coverage` directory in build.

File: cmake/Modules/CodeCoverage.cmake (hunk 3 above)
Comment (knorth55) [reply cmake/Modules/CodeCoverage.cmake]: i change to use `add_custom_command`

File: cmake/Modules/CodeCoverage.cmake [hunk 5]
```diff
@@ -167,17 +162,21 @@ function(ADD_CODE_COVERAGE)
...
+        COMMAND ${LCOV_PATH} --extract ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.removed "'*/${PROJECT_NAME}/*'" --output-file ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.cleaned || echo "WARNING: Not cpp report to output"
+        COMMAND ${CMAKE_COMMAND} -E remove ${PROJECT_BINARY_DIR}/${Coverage_NAME}.base ${PROJECT_BINARY_DIR}/${Coverage_NAME}.total || echo "WARNING: Not cpp report to output"
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/cpp_coverage || echo "WARNING: Error to create cpp coverage dir" || echo "WARNING: Error to create cpp coverage dir"
+        COMMAND cp ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.cleaned ${PROJECT_BINARY_DIR}/${Coverage_NAME}_cpp.info || echo "WARNING: Not cpp report to copy"
+        COMMAND mv ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info ${PROJECT_BINARY_DIR}/cpp_coverage || echo "WARNING: Not cpp report to move"
+        COMMAND mv ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.removed ${PROJECT_BINARY_DIR}/cpp_coverage || echo "WARNING: Not cpp report to move"
+        COMMAND mv ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.cleaned ${PROJECT_BINARY_DIR}/cpp_coverage || echo "WARNING: Not cpp report to move"
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:179]: Can you use `${CMAKE_COMMAND} -E copy`  or `${CMAKE_COMMAND} -E move`? I believe they are more portable than calling the commands directly.

File: cmake/Modules/CodeCoverage.cmake [hunk 6]
```diff
@@ -192,44 +191,48 @@ function(ADD_CODE_COVERAGE)
       set(OMIT_FLAGS --omit ${Coverage_EXCLUDES})
     endif()
 
+    add_custom_command(
+        OUTPUT ${PROJECT_BINARY_DIR}/${Coverage_NAME}_nosetests_python.xml
+        COMMAND ${PYTHON_COVERAGE_PATH} report --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python nosetests report to output"
+        COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_nosetests_python.xml --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python nosetests xml to output"
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/python_nosetests_coverage/ || echo "WARNING: Error to create python nosetests coverage dir"
+        COMMAND mv ${PROJECT_BINARY_DIR}/.coverage* ${PROJECT_BINARY_DIR}/python_nosetests_coverage/ || echo "WARNING: No python nosetests report to move"
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:199]: ditto

File: cmake/Modules/CodeCoverage.cmake [hunk 7]
```diff
@@ -192,44 +191,48 @@ function(ADD_CODE_COVERAGE)
...
         COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_python.xml --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python xml to output"
-        COMMAND ${PYTHON_COVERAGE_PATH} html -d ${Coverage_NAME}_python_html --include "*${REAL_SOURCE_DIR}*" ${OMIT_FLAGS} || echo "WARNING: No python html to output"
-        COMMAND cp ${COVERAGE_DIR}/${Coverage_NAME}_python.xml ${PROJECT_BINARY_DIR} || echo "WARNING: No python xml to copy"
-        COMMAND cp -r ${COVERAGE_DIR}/${Coverage_NAME}_python_html ${PROJECT_BINARY_DIR} || echo "WARNING: No python html to copy"
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${PROJECT_BINARY_DIR}/python_roslaunch_coverage/ || echo "WARNING: Error to create python roslaunch coverage dir"
+        COMMAND cp ${COVERAGE_DIR}/${Coverage_NAME}_python.xml ${PROJECT_BINARY_DIR}/ || echo "WARNING: No python xml to copy"
+        COMMAND cp ${COVERAGE_DIR}/.coverage* ${PROJECT_BINARY_DIR}/python_roslaunch_coverage/ || echo "WARNING: No python coverage to copy"
+        COMMAND cp ${COVERAGE_DIR}/${Coverage_NAME}_python.xml ${PROJECT_BINARY_DIR}/python_roslaunch_coverage || echo "WARNING: No python xml to copy"
```
Comment (garaemon) [inline cmake/Modules/CodeCoverage.cmake:214]: ditto

**Review by garaemon — APPROVED** [review]

File: cmake/Modules/CodeCoverage.cmake (hunk 6 above)
Comment (knorth55) [reply cmake/Modules/CodeCoverage.cmake:199]: For this part, it seems `${CMAKE_COMMAND} -E move` does not support wildcard `*`, which makes command complicated.
I will change the other part.

File: cmake/Modules/CodeCoverage.cmake (hunk 5 above)
Comment (knorth55) [reply cmake/Modules/CodeCoverage.cmake]: I change to use `${CMAKE_COMMAND} -E copy` and `${CMAKE_COMMAND} -E rename`.

**Pull request merged**